client/src/components/Map/components/SuperBlock.js

```diff
@@ -86,14 +86,33 @@ export class SuperBlock extends Component {
   }
 
   render() {
-    const { superBlock, isExpanded, toggleSuperBlock } = this.props;
+    const {
+      superBlock,
+      isExpanded,
+      toggleSuperBlock,
+      nodes,
+      currentChallengeId
+    } = this.props;
+    const blocksForSuperBlock = nodes.filter(
+      node => node.superBlock === superBlock
+    );
+    // If, somehow, the user does not have a current challenge, this opens the
+    // first superBlock.
+    const hasCurrentChallenge = !currentChallengeId
```

Comment:
From what I can gather from this code and your comments, you are trying to determine is the users current challenge is inside this superBlock/block and expanding accordingly.

I can see a couple of pitfalls in this approach, whilst not wrong, they could be considered an anti-pattern.
- We use the `expandedState` to decide which block or superBlock should be expanded. Having individual components decide this is not desired.
- You are essentially using the same code across two components, just working on different data sets. This could and should be extracted out to have only one piece of code to maintain and test.

In the guide navigation, we use the `window.location.pathname` to toggle which parts of the guide navigation should be expanded. I think this is something we could use here also, but it will require some changes in the `get-session-user` api.

There is a utility function in the api that can take a challengeId and return a URI. You could use that api in the `/user/get-session-user` api call and return the resolved path in the `sessionMeta` property of the response.

This handler deals the the session user request:
https://github.com/freeCodeCamp/freeCodeCamp/blob/f687f804d830bdab56b1186ac7d27826ec309d43/api-server/server/boot/user.js#L37-L86

The challengeUrlResolver is here:
https://github.com/freeCodeCamp/freeCodeCamp/blob/f687f804d830bdab56b1186ac7d27826ec309d43/api-server/server/boot/challenge.js#L166-L191

`challengeurlResolver` usage example:

```js
const challengeUrlResolver = createChallengeUrlResolver(app);

const challengeUrl = challengeUrlResolver(`${user.currentChallengeId}`);

challengeUrl //? '/learn/superBlock/block/challenge.dashedName'
```

Example usage with `Observable`'s
```js
// the $ is an rxjs convention to indicate to the reader that this variable is an Observable
const challengeUrlResolver$ = Observable.fromPromise(
  challengeUrlResolver(user.currentChallengeId)
);

// the forkJoin method works like Promise.all, but for Observables
Observable.forJoin(
  foo$,
  bar$,
  baz$,
  (foo, bar, baz) => {
    // foo$, bar$, baz$ Observables have resolved and their output is available
    // in the same order they were passed in
    //
    // return something after doing stuff with foo, bar and baz
  }
)
```


If you store the resolved `challengeUrl` in the `sessionMeta` property of the response, you will then have it available on the client side to toggles superBlocks and blocks as needed, without searching through every challenge for a specific `challengeId`.